torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -129,15 +134,187 @@ def get_tp_parallel_strategy(
+):
+    if job_config.experimental.pipeline_parallel_split_mode == "manual":
+        return pipeline_llama_manual(
+            model, world_mesh, parallel_dims, job_config, device, model_config
+        )
+    elif job_config.experimental.pipeline_parallel_split_mode == "tracer":
+        return pipeline_llama_tracer(
+            model, world_mesh, parallel_dims, job_config, device, model_config
+        )
+    else:
+        raise NotImplementedError(
+            f"{job_config.experimental.pipeline_parallel_split_mode} is not a valid split mode"
+        )
+
+
+def _llama_trace_input(job_config, model_config, device="meta"):
+    """Get meta tensors with the right input shapes used for tracing"""
+    tokens_shape = (job_config.training.batch_size, job_config.training.seq_len)
+    tokens = torch.randint(
+        model_config.vocab_size, tokens_shape, dtype=torch.int64, device=device
+    )
+    return (tokens,)
+
+
+def pipeline_llama_manual(
+    model, world_mesh, parallel_dims, job_config: JobConfig, device, model_config: Dict
+):
+    """
+    This API gets individual torch.nn.Module objects for each pipeline stage (including virtual stages).
+
+    The SPMD parallelisms should be applied to
+    """
+    pp_mesh = world_mesh["pp"]
+    pp_rank = pp_mesh.get_local_rank()
+    pp_size = pp_mesh.size()
+    # heuristically == PP dim but should be a config
+    microbatches = parallel_dims.pp
+    stage_idx = pp_rank  # TODO support virtual stages
+    this_stage_layer_names = split_stage_fqns(
+        _llama_fqns(len(model.layers)),
+        job_config.experimental.pipeline_parallel_split_points,
+        pp_rank,
+    )
+
+    if pp_rank < pp_size - 1:
+        model.norm = None
+        model.output = None
+    if pp_rank > 0:
+        model.tok_embeddings = None
+    names = list(model.layers.keys())
+    for name in names:
+        if f"layers.{name}" not in this_stage_layer_names:
+            del model.layers[name]
+
+    logger.info(f"PP rank {pp_rank} is using this model chunk\n{model}")
+
+    # TODO(whc) once ManualPipelineStage supports lazy shape inference, we can leave model on meta device longer and
+    # get rid of the input shape hardcoded here. For now, it should not be a big deal since we only materialize the
+    # layers of the model that map to this stage, not the whole model.
+
+    if pp_rank == 0:
+        # first layer
+        input = torch.randint(
+            model_config.vocab_size,
+            size=(job_config.training.batch_size, job_config.training.seq_len),
+            dtype=torch.int64,
+            device=device,
+        )
+    else:
+        # later layers (assume all start w/ a transformer layer)
+        input = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=torch.bfloat16 if parallel_dims.dp_enabled else torch.float32,
```

Comment:
I think `dtype=torch.bfloat16 if parallel_dims.dp_enabled` is a bit arbitrary.

Reply:
if your point is that its fragile to do shape inference before we have the full/real model to do inference on, i agree-  but if we put inference into the runtime, I think that would be solved.

or, are you saying there might be a cleaner way to decide this (like adding a config option for runtime dtype, and then checking that instead of just checking 'dp enabled'?)  maybe that's a good idea.

Reply:
Sorry for not being clear enough. Yeah, I was more referring to the latter, i.e. `dp_enabled` != `bfloat16` -- just in case this assumption leads to unexpected effects.

Reply:
Yea I think we need to expose a dtype config and use that for both places.